lib/clean.py

```diff
@@ -691,3 +691,37 @@ def strip_birth_date(df: pd.DataFrame, cols: list[str]) -> pd.DataFrame:
     for col in cols:
         df.loc[:, col] = strip_birth_day_and_month(df[col])
     return df
+
+
+def canonicalize_officers(
+    df: pd.DataFrame,
+    clusters: list[tuple],
+    uid_column: str = "uid",
+    first_name_column: str = "first_name",
+    last_name_column: str = "last_name",
+) -> pd.DataFrame:
+    for cluster in clusters:
+        uid, first_name, last_name = None, "", ""
+        for idx in cluster:
+            row = df.loc[df[uid_column] == idx]
+            first_names = []
+            last_names = []
+            if (
+                uid is None
+                or len(row[first_name_column]) > len(first_name)
+                or (
+                    len(row[first_name_column]) == len(first_name)
+                    and len(row[last_name_column]) > len(last_name)
+                )
+            ):
+                uid = idx
+                first_name = row[first_name_column]
+                first_names.append(first_name)
+
+                last_name = row[last_name_column]
+                last_names.append(last_name)
+
+                df.loc[df[uid_column].isin(cluster), uid_column] = uid
+                df.loc[df[uid_column].isin(cluster), first_name_column] = first_names
+                df.loc[df[uid_column].isin(cluster), last_name_column] = last_names
+    return df
```

Comment:
So it looks like the only thing that changed is that you are assigning a list instead of a string to a column? That's really weird, I swear assigning a single string to an entire column used to work perfectly. Not that there's anything wrong with assigning a list of course. But other than that this code does have problems:
- Since the `first_names` and `last_names` are all reset in each new iteration, they are always going to be a 1-item list at the moment of column assignment.
- The column assignments:
```python
                df.loc[df[uid_column].isin(cluster), uid_column] = uid
                df.loc[df[uid_column].isin(cluster), first_name_column] = first_names
                df.loc[df[uid_column].isin(cluster), last_name_column] = last_names
```
should be done outside of the second for loop. It is very strange that this isn't crashing for any cluster that has more than 1 uid. There is a dimensions miss-match: for every n-item cluster, you're assigning:
- a uid string to an (n x 1) uid series, which should work of course.
- a list of first names which only ever has 1 item (1 x 1) to an (n x 1) first name series, which strangely doesn't crash.  I guess a list of 1 string is treated as a single string?
- (1 x 1) last names to an (n x 1) last name series, which works similarly to the first names.

All in all, I think you should get rid of `first_names` and `last_names` as assigning a single string to a series is possible as shown with the uid assignment. I suspect the previous implementation has a bug that somehow made the `first_name` and `last_name` empty. And please move the column assignments out of the second for loop.

Reply:
The code I pushed seems to have fixed the bug that made the `first_name` and `last_name` empty.

Reply:
What is the purpose of `' '.join(row[last_name_column])`? I thought the name cells are all single strings, not iterables.

Reply:
Why it works? No idea. I was hoping that you could tell me. 

Running 

```
uid = idx
first_name = row[first_name_column]
last_name = row[last_name_column]
```
returns the correct uid, with but does not return the `first_name` and `last_name` for the row(s) whose `uid `has been modified.

Running 

```
uid = idx
first_name = ' '.join(row[first_name_column])
last_name = ' '.join(row[last_name_column])
```
returns the `uid`,  in addition to the `first_name` and ` last_name` for all rows, including the row(s) which has been modified.

Reply:
In other news, I've decided that this test is insufficient after reviewing the cluster sizes in `match/new_orleans_so` i.e., there are 6 individuals who are all likely to be the same person whereas this test has been limited to two individuals.

Reply:
Well, one should always keep all possible scenarios in mind when writing tests and make sure the tests sufficiently address all edge cases. Do you mind if I take a stab at writing this function and its tests?